core/src/Streamly/Internal/Data/MutArray/Type.hs

```diff
@@ -204,28 +219,64 @@ module Streamly.Internal.Data.MutArray.Type
 
     -- ** Growing and Shrinking
     -- | Arrays grow only at the end, though technically it is possible to
-    -- grow on both sides and therefore have a cons as well as snoc. But that
-    -- will require both lower and upper bound in the array representation.
+    -- grow on both sides and therefore we can have a cons as well as snoc. But
+    -- cons is not implemented yet.
 
     -- *** Appending elements
-    , snocWith
+    -- | snoc is the fundamental operation for growing arrays. Streaming folds,
+    -- appending streams can be implemented in terms of snoc.
+
+    -- XXX snoc 128/256/512 bit data using SIMD.
+    , snocWith -- XXX snocGrowWith
     , snoc
-    , snocLinear
+    , snocGrowBy
     , snocMay
     , unsafeSnoc
 
+ -- , revSnoc -- cons
+ -- , revSnocGrowBy  -- consGrowBy
+
+    -- *** Folds for appending streams
+    -- | Fundamentally these are a sequence of snoc operations.
+    -- Folds are named "append" whereas joining two arrays is named as "splice".
+
+    , appendWith -- XXX replace by pure appendGrowWith
+
+    , unsafeAppendMax -- can be renamed to unsafeAppendN later
+    , appendMax -- can be renamed to appendN later
+ -- , appendMin -- like createMinOf, supplies a min hint to reduce allocs
+ -- , appendGrowWith
+    , append2   -- to be renamed to append later
+    , appendGrowBy
+
+ -- , revAppend
+ -- , revAppendN
+ -- , revAppendGrowBy
+
     -- *** Appending streams
-    , unsafeAppendN
-    , appendN
-    , appendWith
-    , append
+    -- | Fundamentally these are a sequence of snoc operations. These are
+    -- convenience operations implemented in terms of folds.
     , unsafeAppendPtrN
-
-    -- *** Appending arrays
-    , spliceCopy
-    , spliceWith
+    , appendPtrN
+    , appendCString#
+ -- , appendStreamGrowWith
+    , appendStream
+    , appendStreamN
+ -- , appendStreamGrowBy
+
+    -- *** Splicing arrays
+    -- | TODO: We can replace memcpy with stream copy using Word64. Arrays are
+    -- aligned on 64-bit boundaries on 64-bit CPUs. A fast way to copy an
+    -- array is to unsafeCast it to Word64, read it as a stream, write the
+    -- stream to Word64 array and unsafeCast it again. We can use SIMD
+    -- read/write as well.
+
+    , spliceCopy -- XXX freeze and splice instead?
     , splice
-    , spliceExp
+    , spliceWith -- XXX spliceGrowWith
```

Comment:
Is `Grow` required in the name.
We're making the name longer.
Isn't `Grow` implicit?

Reply:
spliceWith sounds like splice an array with another - which is also implicit. So spliceGrowWith is more explicit to say that we are specifying a function for grow behavior.